internal/tcp/tcp.go

```diff
@@ -228,349 +189,219 @@ var (
-		uint16(pkt.ECN)<<6 | // 3 bits
-		uint16(pkt.ctrl) // bottom 6 bits
-	binary.Write(buf, binary.BigEndian, mix)
-
-	binary.Write(buf, binary.BigEndian, pkt.window)
-	binary.Write(buf, binary.BigEndian, pkt.checksum)
-	binary.Write(buf, binary.BigEndian, pkt.urgent)
-
-	for _, option := range pkt.options {
-		binary.Write(buf, binary.BigEndian, option.Kind)
-		if option.Length > 1 {
-			binary.Write(buf, binary.BigEndian, option.Length)
-			binary.Write(buf, binary.BigEndian, option.Data)
+	// When replying with SYN+ACK, a time-stamped payload is included
+	if flags.syn != false && flags.ack != false {
+		payloadTime, err := timeNow().MarshalBinary()
+		if err != nil {
+			return nil, err
 		}
+		payloadLayer := gopacket.Payload(payloadTime)
+		payloadLayer.SerializeTo(buf, optsTCP)
 	}
 
-	if pkt.tcpPayload != (tcpPayload{}) {
-		byteTime, err := pkt.tcpPayload.Ts.MarshalBinary()
-		if err == nil {
-			binary.Write(buf, binary.BigEndian, byteTime)
-		}
+	tcpLayer := &layers.TCP{
+		SrcPort:    layers.TCPPort(srcPort),
+		DstPort:    layers.TCPPort(dstPort),
+		Seq:        seqNum,
+		Ack:        ackNum,
+		DataOffset: uint8(tcpHdrSize / 4), // TCP Header size in 32-bit words
+		SYN:        flags.syn,
+		RST:        flags.rst,
+		ACK:        flags.ack,
+		Window:     defines.TCPWindowSize,
+		Checksum:   0, // computed upon serialization
 	}
 
-	out := buf.Bytes()
-
-	return out
-}
+	// Length of TCP portion is payload length + fixed 20 bytes for Header
+	tcpLen := tcpHdrSize + len(buf.Bytes())
 
-func (tcp *tcpHeader) hasFlag(flagBit byte) bool {
-	return tcp.ctrl&flagBit != 0
-}
-
-// TCP Checksum
-func checksum(af string, data []byte, srcip, dstip *net.IP) (uint16, error) {
-
-	// the pseudo header used for TCP c-sum computation
-	var pseudoHeader []byte
-
-	pseudoHeader = append(pseudoHeader, *srcip...)
-	pseudoHeader = append(pseudoHeader, *dstip...)
-	switch af {
-	case "ip4":
-		pseudoHeader = append(pseudoHeader, []byte{
-			0,
-			6,                  // protocol number for TCP
-			0, byte(len(data)), // TCP length (16 bits), w/o pseudoheader
-		}...)
-	case "ip6":
-		pseudoHeader = append(pseudoHeader, []byte{
-			0, 0, 0, byte(len(data)), // TCP length (32 bits), w/0 pseudoheader
-			0, 0, 0,
-			6, // protocol number for TCP
-		}...)
-	default:
-		return 0, errors.New("unhandled AF family")
+	ipLayer, err := ip.GetIPHeaderLayer(af, uint8(dscpv), uint16(tcpLen), srcAddr, dstAddr)
+	if err != nil {
+		return nil, err
 	}
 
-	body := make([]byte, 0, len(pseudoHeader)+len(data))
-	body = append(body, pseudoHeader...)
-	body = append(body, data...)
+	tcpLayer.SetNetworkLayerForChecksum(ipLayer)
 
-	bodyLen := len(body)
-
-	var word uint16
-	var csum uint32
-
-	for i := 0; i+1 < bodyLen; i += 2 {
-		word = uint16(body[i])<<8 | uint16(body[i+1])
-		csum += uint32(word)
+	if err = tcpLayer.SerializeTo(buf, optsTCP); err != nil {
+		return nil, err
 	}
 
-	if bodyLen%2 != 0 {
-		csum += uint32(body[len(body)-1])
+	switch ipLayer.(type) {
```

Comment:
when you're using the value in the switch there is a short-cut

```go
switch layer := ipLayer.(type) {
case *layers.IPv4:
  err = layer.SerializeTo(buf, optsIP)
case *layers.IPv6:
  err = layer.SerializeTo(buf, optsIP)
}
```
https://golang.org/ref/spec#TypeSwitchStmt